resistor-colors: New exercise implementation

pgaspar opened the pull request:
Closes #938. The goal was to have something that passes the tests from https://github.com/exercism/problem-specifications/pull/1466 so we can start the conversation.

TODO:
- [x] Add to config.json

F3PiX commented:
Aha cool! That helps!

This was supposed  to be an array exercise, making use of the fun fact that the colours are coded 0..9, so we can use the index in the array to find the code. 
Ideally, the exercise would steer into the direction of something like:

```
color_bands.map {|color| COLORS.index(color) }.join.to_i
```
Can we make that happen?

pgaspar commented:
I assume `color_bands` is the input? If so you should suggest changing the problem specification so it defines the method input as an array and not two separate attributes.

> Can we make that happen?

Other than that, I can change the solution to be more in line with your snippet. If the input doesn't change to an array, though, I expect many students to prefer something along the lines of:

```ruby
"#{COLORS.index(color1)}#{COLORS.index(color2))".to_i
```

vs.

```ruby
[color1, color2].map {|color| COLORS.index(color) }.join.to_i
```

This last solution does a lot more work, specially for something that will only ever receive 2 colors (right?). It makes an array out of the two color variables, maps through it, joins the result creating a string, converts the string back to integer. The first approach makes a string using interpolation and converts it to integer.

F3PiX commented:
The input _is_ going to change to an array. To be continued.

F3PiX commented on exercises/resistor-significants/.meta/solutions/resistor_significants.rb:
```diff
@@ -0,0 +1,21 @@
+    violet
+    grey
+    white
+  )
```
Let's freeze this.

F3PiX commented:
> This [map] solution does a lot more work, specially for something that will only ever receive 2 colors (right?). It makes an array out of the two color variables, maps through it, joins the result creating a string, converts the string back to integer. The first approach makes a string using interpolation and converts it to integer.

- A resistor can have more than 2 bands (1 upto 6), and students can be well aware of that fact through the description. But, in this exercise it is always about 2 bands, yes.
- I'd think that the `map` solution is better: more idiomatic, more open/close, better readability, better maintainability. And given that we don't care about optimization at this point in the track. 
Is it true then that the map solution is better? 
- If ^ is  true, we can go two ways: when students create a solution with string interpolation, the mentoring can point to `map`. Or, we just accept the string interpolation as 'minimal solution for approval', and use the next exercise in the series, that adds the third band, to illustrate the usefulness of een open/close solution. 
- If ^^ is not true, we just go with the string interpolation.

pgaspar commented:
> * I'd think that the `map` solution is better: more idiomatic, more open/close, better readability, better maintainability. And given that we don't care about optimization at this point in the track.
  Is it true then that the map solution is better?

I think this is all a bit arguable. If the input is going to be an array from the get go I'd do something like this, just because I prefer mapping over accessing array elements directly using `[0]` or `first`. RE open/closed, what scenario are you thinking about specifically?

So, to clarify, I agree with changing the solution when the input changes on the problem specification now that the input is an array. But I still have some doubts regarding making it about open/closed - a good example of how extension and modification would look like here would help!

pgaspar commented:
The exercise has been renamed from `resistor-significants` to `resistor-colors`. I'm updating the PR to reflect that.

pgaspar commented:
The build is failing because `config.json` was not updated. @F3PiX if you can add a commit with that change I'd appreciate it!

pgaspar commented:
Squashed everything into a single commit. Squashed commits:

- resistor-significants: Initial approach
- resistor-significants: map approach
- resistor-significants: Array input
- resistor-colors: New exercise

pgaspar commented:
Looks good to me @F3PiX. Do you want someone else to review this?

F3PiX commented:
Thanks Pedro :bow: I added the config.json.

As stated in the OP, we should discuss a bit more how we want to use the exercise in the track. 

- The most important thing: avoid that the student has to decide on the data structure. I'd prefer to give them the COLORS array, by:
  - Adding a hint to the Readme, where they can copy/paste the COLORS array 
  OR  
  - Adding a stub file with the COLORS array in it. Discussed [here](https://github.com/exercism/ruby/issues/541). I'm not a fan of adding stub files everywhere, but in this case, giving them the data structure for the Colors is a big pro to me.   

The other thing on my wish list: add some script exercises. This could be a candidate for the first,  simple one.  Using `gets.chomp` is part of most beginner tutorials I know of, but `ARGV` would be more interesting, but also levels it up, and in that case the logic may be a bit too simple. 

Thoughts?

pgaspar commented:
> The most important thing: avoid that the student has to decide on the data structure. I'd prefer to give them the COLORS array

I'm not entirely sure why we want to force people to follow this specific approach so much. I thought Exercism was not about prescribing the single correct solution but allowing for experimentation and leveraging the fact that we have real mentors to explore the pros and cons of other approaches.

> The other thing on my wish list: add some script exercises. This could be a candidate for the first, simple one. Using gets.chomp is part of most beginner tutorials I know of, but ARGV would be more interesting, but also levels it up, and in that case the logic may be a bit too simple.

I shared some of my thought regarding this on Slack, but I'll share them here too for others to read: This seems to deviate a bit from the test-first approach. Throughout the track we try to get people to run their code through their tests, not by calling it via terminal. In `TwoFer` it's common to see people use `puts` and `p` and to have `TwoFer.two_fer("Some String")` at the end of the file. That is usually indicative that they’re not running the tests but rather manually debugging through their console / IDE.

Another point is that `gets.chomp` is hard to test. Here are [some examples](https://stackoverflow.com/questions/16948645/how-do-i-test-a-function-with-gets-chomp-in-it) of what people do. This by itself forces the code to have a certain structure which may be confusing for beginners.

I'm skeptical about the need to have script exercises in Exercism. I'd rather if we continued focusing on running exercises via the tests, which is a good habit to interiorize from early on IMO.

Curious about what others think.

F3PiX commented:
@pgaspar My conclusion for now is: let's merge it as is, and I'll monitor what happens.  
 
re:  the scripting exercises: they shouldn't be non-TDD. And I totally agree that an exercise at this level should have easy to read tests! Good point.

kotp commented on exercises/resistor-colors/README.md:
```diff
@@ -0,0 +1,52 @@
+
+The band colors are encoded as follows:
+
+- Black: 0
```
Traditionally, these bands are expressed as starting with value 1.  It fits the mnemonic that we use to remember the values.  Since this describes the problem domain, there is no reason to order them as we order things in the programming domain (often using 0 as the starting point).

F3PiX replied:
That's not what I found.

For quick reference:
https://en.wikipedia.org/wiki/List_of_electronic_color_code_mnemonics
All these mnemonics starts with BB, for Black and Brown.

In Dutch it is `**Z**ij Bracht Rozen op Gerrits Graf bij view grauw weer`
Z for Zwart = Black. `weer` = Wit = White

The fun is: it's one of the rare real life cases where the index 0 _does_ have meaning, and that was the inspiration for the exercise.  :joy:

kotp replied:
Well, it has been quite a few decades. :)  And I am going off of my rusty memory.  Good to go then.

F3PiX commented:
Thanks all, and especially @pgaspar for delivering _yet another_ nice addition to the track! 

🍾

pgaspar commented:
@kotp I believe merging is blocked until you ✅ the PR.

kotp commented:
I noted that it is currently not core and unlocked by null.

F3PiX commented:
Thanks Victor!
